packages/component/src/types.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ComponentContainer } from './component_container';
+
+export const enum InstantiationMode {
+  LAZY = 'LAZY', // Currently all components are LAZY in JS SDK
+  EAGER = 'EAGER'
+}
+
+/**
+ * PUBLIC: A public component provides a set of public APIs to customers. A service namespace will be patched
+ * onto `firebase` namespace. Assume the component name is `test`, customers will be able
+ * to get the service by calling `firebase.test()` or `app.test()` where `app` is a `FirebaseApp` instance.
+ *
+ * PRIVATE: A private component provides a set of private APIs that are used internally by other
+ * Firebase SDKs. No serivce namespace is created in `firebase` namespace and customers have no way to get them.
+ */
+export enum ComponentType {
+  PUBLIC = 'PUBLIC',
+  PRIVATE = 'PRIVARE'
+}
+
+/**
+ * Factory to create an instance of type T, given a ComponentContainer.
+ * ComponentContainer is the IOC container that provides {@link Provider}
+ * for dependencies.
+ *
+ * NOTE: The container only provides {@link Provider} rather than the actual instances of dependencies.
+ * It is useful for lazily loaded dependencies and optional dependencies.
+ */
+export type InstanceFactory<T = unknown> = (
```

Comment:
IIUC, return only generic only applies to the concrete function declaration.
Here It's just a generic type to help create concrete types which are in turn used to type functions.

Reply:
Yeah that's true, I just realized how it's actually used. :thinking: :sweat_smile: 

You could make this `InstanceFactory<T extends Service>` if you do `NameServicePairs`, but if not, this is fine too.

Reply:
I think we can keep this type independent of  the `NameServicePairs`. To make everything 100% type safe, we'd need to tie `Component` with `NameServicePairs` as well. I feel it might be too restricting because people should be able to use `Component` without doing extra things.

Reply:
Isn't a `Component` supposed to be implementing a single service? It feels like a perfect match to me, you could define `Component` as `export class Component<T extends Name> {`, and still all a service would need to do is to extend the `NameServicePairs` interface, which they're already doing AFAICT. But TBH I might not be aware of all the use cases of `Component`.